lib/elixir_analyzer/submission.ex

```diff
@@ -8,130 +8,129 @@ defmodule ElixirAnalyzer.Submission do
-    do_prepend_comment(r, comment)
+  defp get_summary(submission = %__MODULE__{halted: true, comments: comments})
+       when length(comments) == 0 do
+    case submission.halt_reason do
+      nil -> "Analysis was halted."
+      _ -> "Analysis was halted. #{submission.halt_reason}"
+    end
   end
 
-  def prepend_comment(r = %__MODULE__{}, {comment}) when is_binary(comment) do
-    do_prepend_comment(r, %{"comment" => comment})
-  end
+  @comment_types ~w{celebratory essential actionable informative}a
+  @default_type_frequencies @comment_types |> Enum.map(&{&1, 0}) |> Enum.into(%{})
+
+  defp get_summary(%__MODULE__{comments: comments}) do
+    type_frequencies_in_comments = Enum.frequencies_by(comments, fn comment -> comment.type end)
+
+    summary =
+      @default_type_frequencies
+      |> Map.merge(type_frequencies_in_comments)
+      |> build_summary_response()
 
-  def prepend_comment(r = %__MODULE__{}, {comment, params})
-      when is_binary(comment) and is_map(params) do
-    params
-    |> Map.keys()
-    |> Enum.all?(fn key -> Kernel.is_atom(key) or Kernel.is_binary(key) end)
-    |> case do
-      false ->
-        raise ArgumentError, "key type must be atom or binary"
-
-      true ->
-        do_prepend_comment(r, %{"comment" => comment, "params" => params})
+    case summary do
+      nil -> "Submission analyzed. No automated suggestions. Great work! 🚀"
+      _ -> summary
     end
   end
 
-  defp do_prepend_comment(r = %__MODULE__{}, comment) do
-    %{r | comments: [comment | r.comments]}
-  end
+  defp build_summary_response(type_frequencies, keys_to_build \\ @comment_types, acc \\ nil)
+  defp build_summary_response(_type_frequencies, [], acc), do: acc
 
-  def append_comment(r = %__MODULE__{}, comment) when is_binary(comment) do
-    do_append_comment(r, comment)
-  end
+  # Celebratory
+  defp build_summary_response(t = %{celebratory: 0}, [:celebratory | rest], acc),
+    do: build_summary_response(t, rest, acc)
 
-  def append_comment(r = %__MODULE__{}, {comment}) when is_binary(comment) do
-    do_append_comment(r, %{"comment" => comment})
-  end
+  defp build_summary_response(t = %{celebratory: _}, [:celebratory | rest], _acc),
+    do: build_summary_response(t, rest, "🎉")
 
-  def append_comment(r = %__MODULE__{}, {comment, params})
-      when is_binary(comment) and is_map(params) do
-    Map.keys(params)
-    |> Enum.all?(fn key -> Kernel.is_atom(key) or Kernel.is_binary(key) end)
-    |> case do
-      false ->
-        raise ArgumentError, "key type must be atom or binary"
+  # Essential
+  defp build_summary_response(t = %{essential: 0}, [:essential | rest], acc),
+    do: build_summary_response(t, rest, acc)
 
-      true ->
-        do_append_comment(r, %{"comment" => comment, "params" => params})
-    end
-  end
+  defp build_summary_response(_, [:essential | _], acc),
+    do: "#{acc} Check out the comments for things to fix." |> String.trim_leading()
```

Comment:
I see that for essential, actionable, and informative comments, the accumulator is used when producing the summary, but there is no recursive call. For celebratory comments, the accumulator is set to "🎉" and there is a recursive call. Does that mean this is the expected output?

0 celebratory, 1 essential, 0 actionable:
```
"Check out the comments for things to fix."
```

0 celebratory, 1 essential, 1 actionable:
```
"Check out the comments for things to fix."
```

1 celebratory, 1 essential, 1 actionable:
```
"🎉 Check out the comments for things to fix."
```

The last case seems a bit inappropriate.

I think it would be nice to have unit tests that will demonstrate how choosing the summary works with different combinations of different comment types. I agree with the current behavior of the summary when there are essential, actionable, and/or informative comments, but I think for celebratory comments, we should only have a celebratory summary when there are no other types of comments.

Reply:
Ya, I was just trying to give it some sense of personalized message based on the results of the analyzer. The last case doesn't feel right to me either. Will iterate